### PR TITLE
refined functions to get packages indices

### PR DESCRIPTION
**Pull request by rgildein**
Add the _get_package_indices, which is searching all "**/Packages*" in dists, but instead of picking up only the first path, it's filtering a path from the same parent directory.

e.g.
```python
# from
[
  "bionic-backports/universe/binary-amd64/Packages.xz",
  "bionic-backports/universe/binary-amd64/Packages.gz",
  "bionic-backports/universe/binary-amd64/Packages",
  "bionic-backports/main/binary-amd64/Packages.xz",
  "bionic-backports/main/binary-amd64/Packages.gz",
  "bionic-backports/main/binary-amd64/Packages",
]
# to
[
  "bionic-backports/main/binary-amd64/Packages",  
  "bionic-backports/universe/binary-amd64/Packages",
]
```

### CODE REVIEW

**Comment by rgildein** [comment]
After testing the approach with hard links, I come to the conclusion that the `check-packages` and `clean-up-packages` actions could and should be marked as deprecated, since the `synchronize` action is already doing the cleaning and with hard link if you remove the snapshot or update mirror lists (context of `/var/spool/apt-mirror/mirror` will change), all packages will be properly cleaned.

File: src/utils.py [hunk 1]
```diff
@@ -58,6 +54,26 @@ def find_packages_by_indices(indices, base=""):
     return packages
 
 
+def _get_archive_root(pool: Path) -> Path:
+    """Get arive path from pool path."""
```
Comment (Vultaire) [inline src/utils.py:58]: arive -> archive

File: src/utils.py [hunk 2]
```diff
@@ -58,6 +54,26 @@ def find_packages_by_indices(indices, base=""):
...
+    if pool.is_symlink():
+        return pool.readlink().parent.absolute()
+
+    return pool.parent.absolute()
+
+
+def _get_package_indices(dists: Path) -> List[Path]:
+    """Get package indeces."""
```
Comment (Vultaire) [inline src/utils.py:66]: indeces -> indices

**Review by Vultaire — CHANGES_REQUESTED** [review]
The first 2 commits, I'm a +1 once the spelling errors (highlighted) are addressed.

The last 2 commits also look good for functionality from this point forward.  However, it introduces behavior which I'm concerned may break existing snapshots.

I think there is no harm in merging the first 2 of these commits, but we need to be careful regarding the second 2 commits to ensure that files in existing snapshots do not get accidentally dropped due to any automatic cleanups.  Or, if we can't reasonably do so, we may need to consider migration of those snapshots, or perhaps "blocking" the charm if old snapshots are detected so new cleanup/sync actions can't be run until some sort of acknowledgement is received.  ...Just a few thoughts.

File: src/utils.py (hunk 1 above)
Comment (rgildein) [reply src/utils.py:58]: Addressed these typos in [#28](https://github.com/canonical/charm-apt-mirror/pull/28).

**Comment by rgildein** [comment]
I split this PR to 3 different ones:
- #27 to drop bionic support
- #28 to fix bug [lp#2046088](https://bugs.launchpad.net/charm-apt-mirror/+bug/2046088)
- #29 to introduced hard link usage

**Pull request closed without merging**